twisted/conch/ssh/keys.py

```diff
@@ -802,9 +808,13 @@ def fingerprint(self):
 
         @rtype: L{str}
         """
-        return nativeString(
-            b':'.join([binascii.hexlify(x)
-                       for x in iterbytes(md5(self.blob()).digest())]))
+        fingerprint_digest=hash_function(self.blob()).digest()
+        if hash_function==sha256:
+            return nativeString(base64.b64encode(fingerprint_digest))
+        else:
+            return nativeString(
+                b':'.join([binascii.hexlify(x)
+                           for x in iterbytes(fingerprint_digest)]))
```

Comment:
I have two notes here. Firstly, we're missing coverage of the old MD5 branch: we'll need to add a test that covers it and confirms the behaviour.

Secondly, I think this code reveals a bigger problem with this change, which is that `hash_function` is literally a Python callable. While that seems like a neat solution, it means that this function can really quite easily misbehave. For example, if I provide `hashlib.sha512` to the `hash_function`, we will fall into the bottom branch of this code and present it like an MD5 hash, when what we should do is refuse to process.

It would be better, I suspect, to use an enum here. Twisted has support for [constants with values](https://twistedmatrix.com/documents/current/core/howto/constants.html#constants-with-values) in an enumerated form. That would give us the opportunity to do this:

```python
import hashlib
from twisted.python.constants import ValueConstant, Values

class FingerprintHashFunctions(Values):
    """
    Constants representing the supported hash functions for generating
    public key fingerprints.
    """
    MD5 = ValueConstant(hashlib.md5)
    SHA256 = ValueConstant(hashlib.sha256)
```

The method can then optionally accept this enum and fallback to using `FingerprintHasheFunctions.MD5` internally until the deprecation period is over. That way we can ensure that users can't accidentally misuse this function as easily, while also enabling extensibility (users can subclass this class and add extra values to support other hash functions if they want).

That also allows you to separate the act of building the digest from the act of rendering it. For all cases you build the digest identically, then you can just have an if block that says `if hashFunction is FingerprintHashFunctions.MD5` to control the rendering.